File: src/main/java/com/key/oa/util/JwtUtil.java
package com.key.oa.util;

import io.jsonwebtoken.JwtBuilder;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.time.DateUtils;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.stereotype.Component;

import java.nio.charset.StandardCharsets;
import java.util.*;

/**
 * @author 孙强
 * Jwt基本功能的封装
 */
@Data
@Slf4j
@Component
@ConfigurationProperties(prefix = "key-oa.jwt")
public class JwtUtil {
    /**
     * 默认有效时间，默认1天
     */
    private Integer ttlByMinute = 60 * 24;

    /**
     * 证书签发方
     */
    private String issuer = "jyjdal";

    /**
     * 加密时使用的盐
     */
    private String salt;

    public String generate(String subject) {
        return generateWithClaims(subject, null);
    }

    public String generateWithClaims(String subject, Map<String, Object> claims) {
        builder().setSubject(subject);
        if (Objects.isNull(claims)) {
            return builder().compact();
        }
        return builder().setClaims(claims).compact();
    }

    private JwtBuilder builder() {
        return Jwts.builder()
                .setIssuer(issuer)
                .setIssuedAt(new Date())
                // 默认失效时间为1天
                .setExpiration(DateUtils.addMinutes(new Date(), ttlByMinute))
                .setId(UUID.randomUUID().toString())
                .signWith(Keys.hmacShaKeyFor(Base64.getEncoder().encode(salt.getBytes(StandardCharsets.UTF_8))));
    }
}

File: src/main/java/com/key/oa/util/JwtUtil.kt
package com.key.oa.util

import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtBuilder
import io.jsonwebtoken.JwtParser
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.apache.commons.lang3.time.DateUtils
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.security.Key
import java.util.*

private const val ISSUER: String = "jyjdal"
private const val SALT: String = "1Jvq2wYsh3B2Ex9UZlNwNUMpbCILbvQt"

@Component
class JwtUtil {
    fun generate(subject: String, claims: Map<String, Any?>? = null, ttlByMinute: Int = 60 * 24): String {
        return if (claims == null) {
            builder()
        } else {
            builder().setClaims(claims)
        }
            .setSubject(subject)
            .setExpiration(DateUtils.addMinutes(Date(), ttlByMinute))
            .compact()
    }

    fun parse(token: String): Claims = parser().parseClaimsJws(token).body

    private fun builder(): JwtBuilder = Jwts.builder()
        .setIssuer(ISSUER)
        .setIssuedAt(Date())
        .setId(UUID.randomUUID().toString())
        .signWith(key())

    private fun parser(): JwtParser = Jwts.parserBuilder().setSigningKey(key()).build()

    private fun key(): Key = Keys.hmacShaKeyFor(Base64.getEncoder().encode(SALT.toByteArray(StandardCharsets.UTF_8)))
}
